### PR TITLE
chore(sdk): Separate OIDC URL from completion.

### PR DESCRIPTION
**Pull request by pixlwave**
This PR splits up the OIDC auth when using a code into 2 steps. The first to get the URL and the second to perform the login using the code. The main driver for this has been that it would be nice to use OIDC through the FFI where we have no async support (or closure support). Having 2 separate methods makes the API much simpler with this restriction in place.

### CODE REVIEW

File: crates/matrix-sdk/src/oidc.rs [hunk 1]
```diff
@@ -781,6 +851,23 @@ pub(crate) struct OidcData {
     pub(crate) latest_id_token: RwLock<Option<IdToken<'static>>>,
 }
 
+/// Information about the URL that should be presented to authenticate with the
+/// server.
+#[derive(Debug)]
+pub struct OidcAuthenticationUrl {
```
Comment (pixlwave) [inline crates/matrix-sdk/src/oidc.rs]: I'm not 100% sure if this is the correct name for this struct, totally open to suggestions.

File: crates/matrix-sdk/src/oidc.rs [hunk 2]
```diff
@@ -781,6 +851,23 @@ pub(crate) struct OidcData {
...
+    /// The URL that should be presented.
+    pub url: Url,
+    /// Data necessary to validate a response from the endpoint.
+    pub validation_data: AuthorizationValidationData,
```
Comment (pixlwave) [inline crates/matrix-sdk/src/oidc.rs]: It appears to me that this is the only data needed to verify that the configuration is still correct for the login to complete, but if not I'm happy to add other state here too which we could compare before completing the login.

Comment (zecakeh) [reply crates/matrix-sdk/src/oidc.rs]: I don't think we should pass all the validation data to the clients, they should only care about the `state`.

We could store the validation data in a `HashMap` with the key being the `state` (which means clients need to keep the `Oidc` struct around at least until the authorization returns, unless we store it directly in the `Client`).

Then in `authorize_scope_with_authorization_code` we get the `state` from the client (that should be returned by the provider) along with the code, and check first we indeed have the validation data locally.

I don't know if that's clear but I can implement it if you want.

Comment (pixlwave) [reply crates/matrix-sdk/src/oidc.rs]: This was never meant to be a public property, it should have been private so that the clients couldn't access it. That said I had a go at the hashmap today but it was always empty.

> clients need to keep the Oidc struct around at least until the authorization returns

I now understand this comment 🙈

Will take another look at it tomorrow.

File: crates/matrix-sdk/src/oidc.rs [hunk 3]
```diff
@@ -452,7 +427,80 @@ impl Oidc {
...
+        let client_credentials = self.client_credentials().ok_or(OidcError::NotAuthenticated)?;
+
+        let validation_data = self
+            .client
+            .inner
+            .oidc_validation_data
+            .write()
+            .await
+            .remove(&state)
+            .ok_or(OidcError::InvalidState)?;
```
Comment (pixlwave) [inline crates/matrix-sdk/src/oidc.rs:468]: Do you think it would be too presumptive to remove the `state` comparison from the docs given it *should* fail here now? I guess the only place that could go wrong is if you had 2 authorisations in-flight and one returned the state for the other.

Comment (zecakeh) [reply crates/matrix-sdk/src/oidc.rs:468]: You said it all. If the client only allows a single login at a time they shouldn't care about the state, but if there is the possibility that several logins are happening at the same time, they probably need to care about it.

File: crates/matrix-sdk/src/oidc.rs [hunk 4]
```diff
@@ -421,16 +401,11 @@ impl Oidc {
     /// ```
     ///
     /// [`ClientErrorCode`]: matrix_sdk::oidc::types::errors::ClientErrorCode
-    pub async fn authorize_scope_with_authorization_code<F, Fut>(
+    pub async fn url_for_authorize_scope_with_authorization_code(
```
Comment (zecakeh) [inline crates/matrix-sdk/src/oidc.rs]: That's a bit of a mouthful. This should be enough:

```suggestion
    pub async fn url_for_scope_with_authorization_code(
```

File: crates/matrix-sdk/src/oidc.rs [hunk 5]
```diff
@@ -452,7 +427,80 @@ impl Oidc {
...
+    /// // The access and refresh tokens can be persisted either from the response,
+    /// // or from one of the `Client::session_tokens*` methods.
+    ///
+    /// // You can now make any request compatible with the requested scope.
+    /// let _me = client.whoami().await?;
+    /// # anyhow::Ok(()) })
+    /// ```
+    ///
+    /// [`ClientErrorCode`]: matrix_sdk::oidc::types::errors::ClientErrorCode
+    pub async fn authorize_scope_with_authorization_code(
```
Comment (zecakeh) [inline crates/matrix-sdk/src/oidc.rs]: This is probably more appropriate now (though a bit redundant):
```suggestion
    pub async fn finish_authorization_with_authorization_code(
```

File: crates/matrix-sdk/src/oidc.rs [hunk 6]
```diff
@@ -452,7 +427,80 @@ impl Oidc {
...
 
-        let code = (use_authorization_url)(url, validation_data.state.clone()).await?;
+        let state = validation_data.state.clone();
+
+        self.client.inner.oidc_validation_data.write().await.insert(state.clone(), validation_data);
+
+        Ok(OidcAuthenticationUrl { url, state })
+    }
+
+    /// Authorize the given scope by using the Authorization Code flow.
```
Comment (zecakeh) [inline crates/matrix-sdk/src/oidc.rs]:
```suggestion
    /// Finish authorizing a scope by using the Authorization Code flow.
```

File: crates/matrix-sdk/src/oidc.rs [hunk 7]
```diff
@@ -452,7 +427,80 @@ impl Oidc {
...
+    ///
+    /// This should be called after the registered client has been restored with
+    /// [`Oidc::restore_registered_client()`] and the URL returned by
+    /// `url_for_authorize_scope_with_authorization_code` has been presented.
```
Comment (zecakeh) [inline crates/matrix-sdk/src/oidc.rs]: We can skip the part about `restore_registered_client`, and putting the whole name with the type surrounded by brackets will link to the other method in the docs.

Also we want to call it after the user was redirected to the redirect URI.

File: crates/matrix-sdk/src/oidc.rs [hunk 8]
```diff
@@ -452,7 +427,80 @@ impl Oidc {
...
+    ///
+    /// # Arguments
+    ///
+    /// * `code` - The code received as part of the redirect URI when login was
```
Comment (zecakeh) [inline crates/matrix-sdk/src/oidc.rs]:
```suggestion
    /// * `code` - The code received as part of the redirect URI when the authorization was
```

File: crates/matrix-sdk/src/oidc.rs [hunk 9]
```diff
@@ -452,7 +427,80 @@ impl Oidc {
...
+    ///   successful.
+    ///
+    /// * `state` - The unique string that was included in the
+    ///   `OidcAuthenicationUrl` used for authorization.
```
Comment (zecakeh) [inline crates/matrix-sdk/src/oidc.rs]: We want the one that was in the redirect URI here, to be sure.

File: crates/matrix-sdk/src/oidc.rs [hunk 10]
```diff
@@ -452,7 +427,80 @@ impl Oidc {
...
+    ///
+    /// # Example
```
Comment (zecakeh) [inline crates/matrix-sdk/src/oidc.rs]: I don't think we need to repeat the example if there's nothing new in it.

File: crates/matrix-sdk/src/oidc.rs (hunk 3 above)
Comment (zecakeh) [reply crates/matrix-sdk/src/oidc.rs:468]: So to answer your question, we can probably remove it from the example.

File: crates/matrix-sdk/src/oidc.rs [hunk 11]
```diff
@@ -709,25 +741,75 @@ impl OidcLoginBuilder {
...
         let scope = generate_scope(self.device_id)?;
 
-        let response = self
-            .oidc
-            .authorize_scope_with_authorization_code(&scope, redirect_uri, use_authorization_url)
-            .await?;
+        self.oidc.url_for_authorize_scope_with_authorization_code(&scope, redirect_uri).await
+    }
 
+    /// Login via the Authorization Code flow.
```
Comment (zecakeh) [inline crates/matrix-sdk/src/oidc.rs]:
```suggestion
    /// Finish logging in via the Authorization Code flow.
```

File: crates/matrix-sdk/src/oidc.rs [hunk 12]
```diff
@@ -709,25 +741,75 @@ impl OidcLoginBuilder {
...
+    ///
+    /// # Arguments
+    ///
+    /// * `code` - The code received as part of the redirect URI when login was
+    ///   successful.
+    ///
+    /// * `state` - The unique string that was included in the
```
Comment (zecakeh) [inline crates/matrix-sdk/src/oidc.rs]: Same as above, we want the one from the redirect URI

File: crates/matrix-sdk/src/oidc.rs [hunk 13]
```diff
@@ -709,25 +741,75 @@ impl OidcLoginBuilder {
...
+    ///   `OidcAuthenicationUrl` used for login.
+    ///
+    /// This method should be called after presenting the URL returned by
```
Comment (zecakeh) [inline crates/matrix-sdk/src/oidc.rs]: Can you put that above the arguments? Also same comment, it should be called after the user being redirected.

File: crates/matrix-sdk/src/oidc.rs [hunk 14]
```diff
@@ -709,25 +741,75 @@ impl OidcLoginBuilder {
...
+    /// `url_for_login_with_authorization_code`.
+    ///
+    /// # Example
```
Comment (zecakeh) [inline crates/matrix-sdk/src/oidc.rs]: Same, we don't need to repeat the example.

File: crates/matrix-sdk/src/oidc.rs (hunk 1 above)
Comment (zecakeh) [reply crates/matrix-sdk/src/oidc.rs]: Maybe `OidcAuthorizationData`?

OIDC in general uses the term "authorization", not "authentication"

**Review by zecakeh — COMMENTED** [review]
These are mostly docs/naming adjustments. Other than that it looks good.

File: crates/matrix-sdk/src/oidc.rs [hunk 15]
```diff
@@ -452,7 +427,80 @@ impl Oidc {
...
+        state: String,
+    ) -> Result<AccessTokenResponse, OidcError> {
+        let provider_metadata = self.provider_metadata().await?;
+        let client_credentials = self.client_credentials().ok_or(OidcError::NotAuthenticated)?;
+
+        let validation_data = self
+            .client
+            .inner
+            .oidc_validation_data
+            .write()
```
Comment (zecakeh) [inline crates/matrix-sdk/src/oidc.rs]: If we always write and never read, we should use a `Mutex` instead of `RwLock`

File: crates/matrix-sdk/src/oidc.rs [hunk 16]
```diff
@@ -452,7 +427,80 @@ impl Oidc {
...
+    /// // You can now make any request compatible with the requested scope.
+    /// let _me = client.whoami().await?;
+    /// # anyhow::Ok(()) })
+    /// ```
+    ///
+    /// [`ClientErrorCode`]: matrix_sdk::oidc::types::errors::ClientErrorCode
+    pub async fn authorize_scope_with_authorization_code(
+        &self,
+        code: String,
+        state: String,
```
Comment (zecakeh) [inline crates/matrix-sdk/src/oidc.rs:456]: We probably want to have a helper struct here that we can construct from parsing a URI.

I can do it later if you want.

File: crates/matrix-sdk/src/oidc.rs (hunk 1 above)
Comment (pixlwave) [reply crates/matrix-sdk/src/oidc.rs]: Good shout 👍

File: crates/matrix-sdk/src/oidc.rs (hunk 9 above)
Comment (pixlwave) [reply crates/matrix-sdk/src/oidc.rs]: Whoops, yes we do indeed don't we 🤦‍♂️

File: crates/matrix-sdk/src/oidc.rs (hunk 10 above)
Comment (pixlwave) [reply crates/matrix-sdk/src/oidc.rs]: *phew*

File: crates/matrix-sdk/src/oidc.rs (hunk 15 above)
Comment (pixlwave) [reply crates/matrix-sdk/src/oidc.rs]: TIL, thanks!

File: crates/matrix-sdk/src/oidc.rs (hunk 5 above)
Comment (pixlwave) [reply crates/matrix-sdk/src/oidc.rs]: Would `finish_authorization_with_code` work? Maybe it wouldn't be specific enough?

File: crates/matrix-sdk/src/oidc.rs (hunk 3 above)
Comment (pixlwave) [reply crates/matrix-sdk/src/oidc.rs:468]: I've added an extra sentence to say it should be checked in that case.

File: crates/matrix-sdk/src/oidc.rs (hunk 16 above)
Comment (pixlwave) [reply crates/matrix-sdk/src/oidc.rs:456]: Yes please :)

**Comment by pixlwave** [comment]
Ok, updated 👍

Note: I realised I haven't touched the CLI app in this PR. Would you like me to update that too?

File: crates/matrix-sdk/src/oidc.rs (hunk 5 above)
Comment (zecakeh) [reply crates/matrix-sdk/src/oidc.rs]: I'd like to keep "authorization code" since that's the name of the flow

**Comment by zecakeh** [comment]
> Note: I realised I haven't touched the CLI app in this PR. Would you like me to update that too?

Hmm yeah that would be good to make sure it still works. But I can also fix it later when I add the helper struct

File: crates/matrix-sdk/src/oidc.rs [hunk 17]
```diff
@@ -452,7 +425,44 @@ impl Oidc {
             &mut rand::thread_rng(),
         )?;
 
-        let code = (use_authorization_url)(url, validation_data.state.clone()).await?;
+        let state = validation_data.state.clone();
+
+        self.client.inner.oidc_validation_data.write().await.insert(state.clone(), validation_data);
```
Comment (zecakeh) [inline crates/matrix-sdk/src/oidc.rs]: It should be `lock` not `write`. Does that compile?

File: crates/matrix-sdk/src/oidc.rs [hunk 18]
```diff
@@ -452,7 +425,44 @@ impl Oidc {
...
+        state: String,
+    ) -> Result<AccessTokenResponse, OidcError> {
+        let provider_metadata = self.provider_metadata().await?;
+        let client_credentials = self.client_credentials().ok_or(OidcError::NotAuthenticated)?;
+
+        let validation_data = self
+            .client
+            .inner
+            .oidc_validation_data
+            .write()
```
Comment (zecakeh) [inline crates/matrix-sdk/src/oidc.rs]: Same here

File: crates/matrix-sdk/src/oidc.rs [hunk 19]
```diff
@@ -709,25 +703,40 @@ impl OidcLoginBuilder {
...
+    ///
+    /// * `code` - The code received as part of the redirect URI when login was
+    ///   successful.
+    ///
+    /// * `state` - The unique string received as part of the redirect URI when
+    ///   login was successful
+    ///
+    /// [`ClientErrorCode`]: matrix_sdk::oidc::types::errors::ClientErrorCode
+    #[instrument(target = "matrix_sdk::client", skip_all)]
+    pub async fn login_with_authorization_code(
```
Comment (zecakeh) [inline crates/matrix-sdk/src/oidc.rs]: Didn't I leave a comment to rename this?

```suggestion
    pub async fn finish_login_with_authorization_code(
```

File: crates/matrix-sdk/src/oidc.rs [hunk 20]
```diff
@@ -781,6 +790,15 @@ pub(crate) struct OidcData {
     pub(crate) latest_id_token: RwLock<Option<IdToken<'static>>>,
 }
 
+/// The data needed to perform authorization using OpenID Connect.
+#[derive(Debug)]
```
Comment (zecakeh) [inline crates/matrix-sdk/src/oidc.rs]:
```suggestion
#[derive(Debug, Clone)]
```
There's no harm in allowing to clone the struct and it can be useful for users.

**Comment by pixlwave** [comment]
Ok, updated - and I just tried it out for the first time. It's very nice 👌

File: crates/matrix-sdk/src/oidc.rs (hunk 17 above)
Comment (pixlwave) [reply crates/matrix-sdk/src/oidc.rs]: My bad, I noticed this with the CLI update - I wasn't compiling with the feature enabled 🤦‍♂️

File: crates/matrix-sdk/src/oidc.rs (hunk 19 above)
Comment (pixlwave) [reply crates/matrix-sdk/src/oidc.rs]: Oh, hmmm. I didn't see that one, but I updated the docs for it.

File: crates/matrix-sdk/src/oidc.rs (hunk 20 above)
Comment (pixlwave) [reply crates/matrix-sdk/src/oidc.rs]: Thanks 👍. Intuitively when to add the various `derives` is still something I still haven't got into my head yet.

**Comment by pixlwave** [comment]
And those few comments are fixed too.

File: crates/matrix-sdk/src/oidc.rs [hunk 21]
```diff
@@ -365,25 +364,10 @@ impl Oidc {
...
-    ///   receive instead an `error` with a [`ClientErrorCode`], the `state` and
-    ///   optionally `error_description` and `error_uri` parameters.
-    ///
-    ///   If the callback returns an error, it will be forwarded.
-    ///
     /// # Example
     ///
     /// ```no_run
     /// use matrix_sdk::{Client, Error};
-    /// # use futures::executor::block_on;
```
Comment (zecakeh) [inline crates/matrix-sdk/src/oidc.rs]: Why was this line removed?

File: crates/matrix-sdk/src/oidc.rs [hunk 22]
```diff
@@ -350,7 +348,8 @@ impl Oidc {
         OidcLoginBuilder::new(self.clone())
     }
 
-    /// Authorize the given scope by using the Authorization Code flow.
+    /// Get the URL that should be presented to authorize a given scope with the
```
Comment (zecakeh) [inline crates/matrix-sdk/src/oidc.rs:351]: Sorry I forgot to add a comment earlier, we need to add a sentence in those docs to call the finish method when the user is redirected, since this method is useless without the other.

File: crates/matrix-sdk/src/oidc.rs [hunk 23]
```diff
@@ -646,57 +656,41 @@ impl OidcLoginBuilder {
         self
     }
 
-    /// Login via the Authorization Code flow.
+    /// Get the URL that should be presented to login via the Authorization Code
```
Comment (zecakeh) [inline crates/matrix-sdk/src/oidc.rs:662]: Same here, we need a sentence to say to use the finish method after.

**Review by zecakeh — COMMENTED** [review]
Sorry, more doc fixes :sweat_smile:

File: crates/matrix-sdk/src/oidc.rs (hunk 21 above)
Comment (pixlwave) [reply crates/matrix-sdk/src/oidc.rs]: Honestly, probably just naivety. I assumed that without the async closure, the example didn't use futures any more so it wasn't needed.

Comment (pixlwave) [reply crates/matrix-sdk/src/oidc.rs]: And now I see the `block_on` below.
